File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3013_divide-an-array-into-subarrays-with-minimum-cost-ii/solutions/java/solution.java
import java.util.Arrays;

class Solution {
    public long minimumCost(int[] nums, int k, int dist) {
        int n = nums.length;
        long[][] dp = new long[k][n];
        long[] prefixSum = new long[n + 1];
        
        // Compute prefix sums for efficient range sum calculation
        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + nums[i];
        }
        
        // Base case initialization: the cost of using the first subarray ending at each position
        for (int j = 0; j < n; j++) {
            dp[0][j] = nums[0];
        }
        
        // Fill dp table
        for (int i = 1; i < k; i++) {
            // deque to maintain the minimum values in a sliding window
            Deque<Integer> deque = new LinkedList<>();
            deque.add(i - 1);
            
            for (int j = i; j < n; j++) {
                // Ensure the deque only has relevant indices for the current dp row computation
                while (!deque.isEmpty() && deque.peekFirst() < j - dist - 1) {
                    deque.pollFirst();
                }
                
                // Compute dp[i][j]
                dp[i][j] = dp[i - 1][deque.peekFirst()] + nums[j];
                
                // Update deque to maintain the minimum dp value in the current valid range
                while (!deque.isEmpty() && dp[i - 1][deque.peekLast()] >= dp[i - 1][j]) {
                    deque.pollLast();
                }
                deque.add(j);
                
                // Move to the next j
            }
        }
        
        // Find minimum cost of forming k subarrays
        long minCost = Long.MAX_VALUE;
        for (int start = k - 1; start <= n - 1; start++) {
            if (n - start - 1 <= dist) {
                minCost = Math.min(minCost, dp[k - 1][start]);
            }
        }
        
        return minCost;
    }
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3013_divide-an-array-into-subarrays-with-minimum-cost-ii/solutions/cpp/solution.cpp
#include <vector>
#include <limits>

using namespace std;

class Solution {
public:
    long long minimumCost(vector<int>& nums, int k, int dist) {
        int n = nums.size();
        vector<vector<long long>> dp(k, vector<long long>(n, LLONG_MAX));
        vector<long long> prefixSum(n + 1, 0);
        
        // Compute prefix sums to speed up range sum calculation
        for (int i = 0; i < n; ++i) {
            prefixSum[i + 1] = prefixSum[i] + nums[i];
        }

        // Base case: one partition
        for (int j = 0; j < n; ++j) {
            dp[0][j] = nums[0];
        }

        // Fill dp for all partitions
        for (int p = 1; p < k; ++p) {
            for (int i = p; i < n; ++i) {
                int limit = min(n - k + p + 1, i + dist + 1);
                for (int j = max(p, i - dist); j < limit; ++j) {
                    dp[p][j] = min(dp[p][j], dp[p - 1][i] + nums[j]);
                }
            }
        }

        // Compute the minimum cost from valid partition positions
        long long result = LLONG_MAX;
        int start = max(k - 1, n - dist - 1);
        for (int i = start; i < n; ++i) {
            result = min(result, dp[k - 1][i]);
        }

        return result;
    }
};
